### PR TITLE
feat: no more generated attribute types in CFN layer (L1)

### LINKED ISSUE #1406: Java: Provide a Bill of Materials (BOM) for CDK dependencies
To ease Maven dependency management, provide a BOM for the CDK dependencies.

### PR DESCRIPTION
**Pull request by eladb**
Now that we can represent attributes as native strings and string lists,
this covers the majority of resource attributes in CloudFormation.

This change:

* String attributes are represented as `string` (like before).
* String list attribute are now represented as `string[]`.
* Any other attribute types are represented as `cdk.Token`.

Attributes that are represented as Tokens as of this change:

* amazonmq has a bunch of `Integer` attributes (will be solved by #1455)
* iot1click has a bunch of `Boolean` attributes
* cloudformation has a JSON attribute
* That's all.

A few improvements to cfn2ts:

* Auto-detect cfn2ts scope from package.json so it is more self-contained
  and doesn't rely on cdk-build-tools to run.
* Added a "cfn2ts" npm script to all modules so it is now possible
  to regenerate all L1 via "lerna run cfn2ts".
* Removed the premature optimization for avoiding code regeneration
  (it saved about 0.5ms).

Fixes #1406

BREAKING CHANGE: any `CfnXxx` resource attributes that represented a list of strings are now typed as `string[]`s (via #1144). Attributes that represent strings, are still typed as `string` (#712) and all other attribute types are represented as `cdk.Token`. 

----

### Pull Request Checklist

* [x] Testing
  - Unit test added
  - __CLI change?:__ manually run [integration tests](https://github.com/awslabs/aws-cdk/blob/master/packages/aws-cdk/integ-tests/test.sh) and paste output as a PR comment
  - __cdk-init template change?:__ coordinated update of integration tests with team
* [x] Docs
  - __jsdocs__: All public APIs documented
  - __README__: README and/or documentation topic updated
* [x] Title and Description
  - __Change type__: title prefixed with **fix**, **feat** will appear in changelog
  - __Title__: use lower-case and doesn't end with a period
  - __Breaking?__: last paragraph: "BREAKING CHANGE: <describe what changed + link for details>"
  - __Issues__: Indicate issues fixed via: "**Fixes #xxx**" or "**Closes #xxx**"
* [x] Sensitive Modules (requires 2 PR approvers)
  - IAM Policy Document (in @aws-cdk/aws-iam)
  - EC2 Security Groups and ACLs (in @aws-cdk/aws-ec2)
  - Grant APIs (only if not based on official documentation with a reference)

----

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license.

### CODE REVIEW

File: tools/cfn2ts/lib/codegen.ts [hunk 1]
```diff
@@ -74,7 +74,7 @@ export default class CodeGenerator {
       const legacyResourceName = genspec.CodeName.forLegacyResource(cfnName);
       this.code.line();
 
-      const attributeTypes = this.emitResourceType(resourceName, resourceType);
```
[the remark this answers is not in the document]
Comment (eladb) [reply tools/cfn2ts/lib/codegen.ts]: It doesn't, but added explicit "void" to ensure completeness of refactor.

File: tools/cfn2ts/lib/codegen.ts [hunk 2]
```diff
@@ -286,14 +279,12 @@ export default class CodeGenerator {
...
-      if (at.attributeType.isPrimitive) {
-        if (at.attributeType.typeName.className === 'string') {
-          this.code.line(`this.${at.propertyName} = ${at.constructorArguments}.toString();`);
-        } else {
-          throw new Error(`Unsupported primitive attribute type ${at.attributeType.typeName.className}`);
-        }
-      } else {
-        this.code.line(`this.${at.propertyName} = new ${at.attributeType.typeName.className}(${at.constructorArguments});`);
+      if (at.attributeType === 'string') {
+        this.code.line(`this.${at.propertyName} = ${at.constructorArguments}.toString();`);
```
[the remark this answers is not in the document]
Comment (eladb) [reply tools/cfn2ts/lib/codegen.ts]: I think encoding the type info in L1 is useful for L2 authors (and anyone who uses L1 directly). Granted, there are about 6 attributes in the whole surface area that are now untyped (I can add some docstrings about that), but all the rest are strongly-meta-typed, and align with our approach for modeling tokens.

File: packages/@aws-cdk/alexa-ask/package.json [hunk 3]
```diff
@@ -36,7 +36,8 @@
     "pkglint": "pkglint -f",
     "test": "cdk-test",
     "watch": "cdk-watch",
-    "awslint": "cdk-awslint"
+    "awslint": "cdk-awslint",
+    "cfn2ts": "cfn2ts"
```
[the remark this answers is not in the document]
Comment (eladb) [reply packages/@aws-cdk/alexa-ask/package.json]: I don't see a value in making this hard to discover/use. The modeling of the phases is somewhat arbitrary (I would actually argue this is a "pre-build" step), and I think a simple flat set of scripts is more useful than a specific mental model that's likely not to be the same for everyone.

**Comment by RomainMuller** [comment]
The `Lazy` concept described above... I'm afraid this won't translate super well in all languages (that's a little JSII-hostile). You can't really express `Lazy.stringValue(() => someThing())` in Java for example, at least not in a way that'll achieve the desired outcome over JSII.

**Comment by eladb** [comment]
@RomainMuller I agree, but:
1) It's not worse than the current case (`new Token(()=>...)`)
2) We should model it in a way that allows languages like Java to implement an interface instead of a lambda

**Pull request merged**